cmd/wait-for-files-to-exist/main.go

```diff
@@ -35,85 +36,120 @@ func main() {
+		printOutput("waiting for file %q to exist", f)
+		go func(path string) {
+			defer wg.Done()
+			resultCh <- fileResult{file: path, err: waitForFile(ctx, path)}
+		}(f)
+	}
+
+	go func() {
+		wg.Wait()
+		close(resultCh)
+	}()
+
+	for res := range resultCh {
+		if res.err != nil {
+			exit("error waiting for file %q: %w", res.file, res.err)
+		}
+		printOutput("file %q exists!", res.file)
+	}
+}
+
+func validateFileToWaitFor(fileToWaitFor string) error {
 	st, err := os.Lstat(fileToWaitFor)
 
 	switch {
 	case os.IsNotExist(err):
 		// File does not exist, so we need to watch the parent directory for the file creation.
 		// Ensure to watch the real path rather than a symlinked path which does not work properly with
 		// inotify.
-		dirPath, err := filepath.EvalSymlinks(filepath.Dir(fileToWaitFor))
+		_, err = filepath.EvalSymlinks(filepath.Dir(fileToWaitFor))
 		if err != nil {
-			exit(
-				"failed to evaluate any symlinks to read real directory for %q: %v",
+			return fmt.Errorf(
+				"failed to evaluate any symlinks to read real directory for %q: %w",
 				fileToWaitFor,
 				err,
 			)
 		}
+	case err != nil:
+		return fmt.Errorf("failed to stat %q: %w", fileToWaitFor, err)
+	case st.IsDir():
+		return fmt.Errorf("%q is a directory, not a file", fileToWaitFor)
+	case !st.Mode().IsRegular():
+		return fmt.Errorf("%q exists but is not a regular file - type is %s", fileToWaitFor, st.Mode().String())
+	}
 
-		if dirPath != filepath.Dir(fileToWaitFor) {
-			printOutput(
-				"watching for file %q (real path %q) to exist",
-				fileToWaitFor,
-				filepath.Join(dirPath, filepath.Base(fileToWaitFor)),
-			)
-		} else {
-			printOutput(
-				"watching for file %q to exist",
-				fileToWaitFor,
-			)
+	// File already exists and is a regular file.
+	return nil
+}
+
+func waitForFile(ctx context.Context, fileToWaitFor string) error {
+	w, err := fsnotify.NewWatcher()
+	if err != nil {
+		return fmt.Errorf("creating a new watcher: %w", err)
+	}
+	defer w.Close()
+
+	errChan := make(chan error)
+	go fileLoop(ctx, w, fileToWaitFor, errChan)
+
+	_, err = os.Lstat(fileToWaitFor)
+	if os.IsNotExist(err) {
+		dirPath, err := filepath.EvalSymlinks(filepath.Dir(fileToWaitFor))
+		if err != nil {
+			return fmt.Errorf("failed to evaluate any symlinks to read real directory for %q: %w", fileToWaitFor, err)
 		}
 
 		if err := w.Add(dirPath); err != nil {
-			exit("failed to add watch %q: %v", fileToWaitFor, err)
+			return fmt.Errorf("failed to add watch %q: %w", fileToWaitFor, err)
 		}
-	case err != nil:
-		exit("failed to stat %q: %v", fileToWaitFor, err)
-	case st.IsDir():
-		exit("%q is a directory, not a file", fileToWaitFor)
-	case !st.Mode().IsRegular():
-		exit("%q exists but is not a regular file - type is %s", fileToWaitFor, st.Mode().String())
-	default:
-		// File already exists and is a regular file - we are done.
-		printOutput("file already exists!")
-		return
+	} else {
+		// File already exists, no need to wait.
+		return nil
```

Comment:
This will leave the goroutine started on [line 124](https://github.com/mesosphere/mindthegap/pull/895/files#diff-0adbc6cb9b723105a0d7e8e075b2cd0d7396ec97f302622d3cb3aa4afe34eb4fR124) running for no reason. Let's create a child context from the passed in `ctx`, use that in `fileLoop` call, and always cancel it when this function exits.

Reply:
Good catch, thanks!